### PR TITLE
Improve rendering of piers, dams, cutlines 

### PR DESCRIPTION
**Pull request by pastk**
Pier (line) before:
Screenshot_1643920370
After:
Screenshot_1643923326

Weir (line):
Screenshot_1643920397
Screenshot_1643923400

Dam (line):
Screenshot_1643921490
Screenshot_1643922991

Cutline (line):
Screenshot_1643923840
Screenshot_1643926577

### CODE REVIEW

**Comment by biodranik** [comment]
Жирновато выглядят, слишком бросаются в глаза на фоне остального, хотя объекты может не самые критичные.

**Comment by pastk** [comment]
Dam (area):
Screenshot_1643924474
https://www.openstreetmap.org/way/646504678#map=17/40.81677/41.64580

Somehow "area" rules doesn't apply to area `waterway=dam`, though works well for area piers. It seems like the `area=yes` tag is not picked up - is it hardcoded somewhere?
Also dams are rendered from z15- though should appear at z14 (e.g. line `waterway=weir` appears at z14). I've updated the scale index but it didn't help.
Any idea why?

**Comment by pastk** [comment]
> Жирновато выглядят, слишком бросаются в глаза на фоне остального, хотя объекты может не самые критичные.

Для 4 и 6 скринов (там weir и dam) думаю надо opacity и правда уменьшить.

**Comment by pastk** [comment]
> Жирновато выглядят, слишком бросаются в глаза на фоне остального, хотя объекты может не самые критичные.

Проблема была в том, что они на любом зуме тонкой линией рендерелись
Было
Screenshot_1643920382
Я толщину равномерно распределил по зумам, стало:
Screenshot_1643923379

**Comment by vng** [comment]
1. AFAIR It's impossible to make classifier type (and object respectively) line and area at the same time now. So waterway=dam can be lines only (or area only).
2. Check what did you get in drules_proto_clear.txt to debug drawing scale.

File: data/drules_proto.txt [hunk 1]
```diff
@@ -67927,14 +67927,31 @@ cont {
...
     scale: 15
     lines {
-      width: 1.0
+      width: 2.0
       color: 8421504
       priority: 1970
       cap: BUTTCAP
     }
+    area {
+      color: 13224123
```
Comment (vng) [inline data/drules_proto.txt:67952]: Why the color is different from z14?

File: data/drules_proto.txt [hunk 2]
```diff
@@ -67946,10 +67963,26 @@ cont {
   element {
     scale: 16
     lines {
-      width: 1.0
+      width: 5.0
+      color: 10592406
+      dashdot {
+      }
```
Comment (vng) [inline data/drules_proto.txt:67969]: Hm, we should check drules generator, Why empty dashdot ..

File: data/drules_proto_clear.txt [hunk 3]
```diff
@@ -67536,10 +67569,26 @@ cont {
...
+      priority: 1969
+      cap: BUTTCAP
+    }
+    lines {
+      width: 4.5
+      color: 444629120
       priority: 1970
     }
+    area {
+      color: 12763315
```
Comment (vng) [inline data/drules_proto_clear.txt:67585]: And this color is different from previous .. Is it ok?

File: data/styles/clear/include/Basemap.mapcss [hunk 4]
```diff
@@ -578,45 +574,102 @@ area|z15-[amenity=parking]
...
+line|z14-[man_made=cutline],
+{width: 1;color: @green0;opacity: 1;}
+line|z15[man_made=cutline],
+{width: 1.5;}
+line|z16[man_made=cutline],
+{width: 2;}
+line|z17[man_made=cutline],
+{width: 3;}
+line|z18-[man_made=cutline],
+{width: 4.5;color: @green1;}
```
Comment (vng) [inline data/styles/clear/include/Basemap.mapcss:635]: Is it ok green0 before z18 and green1 from z18?

File: data/styles/vehicle/include/Basemap.mapcss [hunk 5]
```diff
@@ -526,48 +522,88 @@ area|z17-[amenity=parking_space]
...
+{width: 6;}
+
+line|z19-[man_made=pier],
+line|z19-[man_made=breakwater],
+line|z19-[man_made=cutline],
+line|z19-[waterway=dam],
+line|z19-[waterway=weir],
+{width: 8;}
+
+line|z15[man_made=cutline],
```
Comment (vng) [inline data/styles/vehicle/include/Basemap.mapcss:567]: z15- to apply color and opacity after? Or remove cultline width above?

File: data/styles/vehicle/include/Basemap.mapcss [hunk 6]
```diff
@@ -526,48 +522,88 @@ area|z17-[amenity=parking_space]
...
 area|z17[building=train_station],
 {fill-color: @building;opacity: 0.55;casing-width: 1;casing-color: @building_border;}
 
 area|z17-[aeroway=terminal],
 area|z18-[building],
 area|z18-[building:part],
+area|z18-[waterway=dam],
 area|z18-[landuse=garages],
 area|z18-[building=train_station]
 {fill-color: @building;opacity: 0.55;casing-width: 1;casing-color: @building_border;}
```
Comment (vng) [inline data/styles/vehicle/include/Basemap.mapcss:609]: Can we combine last 3 blocks? They have equal drawing rules.

File: data/styles/clear/include/Basemap.mapcss [hunk 7]
```diff
@@ -578,45 +574,102 @@ area|z15-[amenity=parking]
...
+{width: 6;}
+
+line|z19-[man_made=pier],
+line|z19-[man_made=breakwater],
+line|z19-[man_made=cutline],
+line|z19-[waterway=dam],
+line|z19-[waterway=weir],
+{width: 8;}
+
+line|z14-[man_made=cutline],
```
Comment (vng) [inline data/styles/clear/include/Basemap.mapcss:626]: If we define cutline here, remove cutline entries for width above.

**Review by vng — CHANGES_REQUESTED** [review]

File: data/styles/vehicle/include/Basemap.mapcss (hunk 6 above)
Comment (vng) [reply data/styles/vehicle/include/Basemap.mapcss:609]: And I suspect that last 2 blocks should define fill-opacity. Combine 3 blocks into one.

File: data/drules_proto.txt (hunk 2 above)
Comment (pastk) [reply data/drules_proto.txt:67969]: yeah there are empty dashdots at some existing elements like `highway-motorway-bridge` z13; and even seemingly conflicting `lines`
```
  15291   element {
  15292     scale: 13
  15293     lines {
  15294       width: 2.8
  15295       color: 16742950
  15296       priority: 1446
  15297       cap: BUTTCAP
  15298     }
  15299     lines {
  15300       width: 6.0
  15301       color: 15592411
  15302       dashdot {
  15303       }
  15304       priority: 1280
  15305       cap: BUTTCAP
  15306     }
  15307     lines {
  15308       width: 6.4
  15309       color: 8421504
  15310       dashdot {
  15311       }
  15312       priority: 1270
  15313       cap: BUTTCAP
  15314     }
```

File: data/styles/clear/include/Basemap.mapcss (hunk 4 above)
Comment (pastk) [reply data/styles/clear/include/Basemap.mapcss:635]: its common for vegetation objects to become darker with more zoom;
however I changed it to @green0; only but 0.6 opacity

**Comment by pastk** [comment]
Fixes the comments above.
Made dams/weirs and cutlines less pronounced:

Screenshot_1643960959

Screenshot_1643960990

**Comment by pastk** [comment]
Also fixes parkings rendering over forests and some other areas:

Before:
Screenshot_1643963094

After:
Screenshot_1643964261

**Review by vng — APPROVED** [review]

**Comment by biodranik** [comment]
Я бы не делал cutline белыми, они похожи на дороги. Пускай будут более зеленоватыми, в осме нормально же нарисовано.

**Comment by pastk** [comment]
Они и не белые, а светло-зелёные с прозрачностью.

Screenshot_1643960959

Толстые белые это как раз дороги, а просеки тонкие. Можно конечно ещё более прозрачными сделать.

**Comment by biodranik** [comment]
А, это меня попутало с усталости, сорри )

**Comment by pastk** [comment]
> 1. ~AFAIR It's impossible to make classifier type (and object respectively) line and area at the same time now. So waterway=dam can be lines only (or area only)~.
>        UPD: The classifier type can be line and area, but the feature of this type can be area (if geometry is closed) or line (otherwise), not together. If the classifier type was line-only before, you should regenerate all MWM (not only index) to see it like area with the new drawing rules.
> 

Hmm... it seems like after maps regeneration it became _both a line and an area_ (otherwise there won't be a label along the line, right?)
https://www.openstreetmap.org/way/646504678#map=17/40.81677/41.64580

dam-line-area

**Pull request merged**